### ISSUE TITLE
GenomicsDBImport [Documentation/Usage]

### ISSUE THREAD

**Issue by nickhsmith** [broadinstitute/gatk#7898]
## Documentation request
I am trying to follow the germline joint variant best practices and am confused by the steps/examples and how it would work using real WGS or WXS data in a parallelized setting.

### Tool(s) or class(es) involved
GenomicsDBImport v.4.2.6.1 (current)

### Description 
As far as I understand it the joint germline variant calling process is like this (imagine 100 samples):
1. Call variants using `Haplotypecaller` using the gVCF output flag for each sample
2. use the multiple gVCFs (1 per sample) and a set of intervals (WGS_intervals.bed as an example) to build a Genomics DB store using `GenomicsDBImport`
3. Use `GenotypeGVCFs` using the output of `GenomicsDBImport` as the input to consolidate the multiple samples into 1 multi-sample vcf

My question comes from the parallelization/interval splitting during step 2. If I parallelize the GenomicsDBImport across each interval. I would end up with ~300 intervals and subsequently, ~300 GenomicsDB directory paths since I am not adding new samples to an existing DB, then the specified output DB path, "Must be an empty or non-existent directory", which will contain the relevant interval calls for the 100 samples. 

Am I supposed to use the 300 directory paths as input into a single `GenotypeGVCFs` call? Or process each of the 300 intervals into 300 multi-sample vcf files (each with 100 samples) and then merge those into a single vcf file using `GatherVcfs` or some other merging tool.

The examples posted and documentation for `GenomicsDBImport` relay the need for intervals to work effectively, and so does [an old broad lecture recording](https://www.youtube.com/watch?v=XrHt5yBlp80&t=1243s). 

Essentially it boils down to when and how to process and merge the same set of samples (100) over the many intervals (300). If I had 300 compute nodes (as an example) I want to parallelize as much of this as possible. so that each node can process an interval set, and at the end of the process I have 1 VCF file with 100 samples covering the entire range of intervals.

I hope that was clear. Please let me know if you need any more info, or if I should be asking somewhere else. Thanks in advance!

----

**Comment by mjohnsonngi** [comment]
The output of GenomicsDBImport is a directory with a specific structure underneath. The required option here, https://gatk.broadinstitute.org/hc/en-us/articles/5358869876891-GenomicsDBImport#--genomicsdb-workspace-path, provides the "name" of this directory, but you really don't need to worry about the underlying structure. When you provide GenotypeGVCFs the variants, https://gatk.broadinstitute.org/hc/en-us/articles/5358906861083-GenotypeGVCFs#--variant, you're providing that directory. The gendb:// is the beginning of a URI that tells GATK how to read/interact with that structure just like the https:// at the beginning of a web URL.

The -L argument for GenomicsDBImport tells GATK which intervals to look at in the provided samples' data. You may need to adjust your --batch-size as you add more samples or use --consolidate. I've found that with large numbers of intervals and samples I need to use --merge-input-intervals or else the bookkeeping files stored in temp space exceed anything anything reasonable. HaplotypeCaller merges intervals by default, but GenotypeGVCFs does not.

In short, your --genomicsdb-workspace-path MY_DB should be the same as your -V gendb://MY_DB.

**Comment by nickhsmith** [comment]
Sorry that's not quite my question. Once I generate 2 directories through genomic import using 2 intervals because it was faster to parallelize them in 2 separate processes. Can I combine or merge these directories before the genotypegvcf step? Or do I use gathervcfs after that to merge the 2 vcf files?

**Comment by mjohnsonngi** [comment]
In that case, I believe you want GatherVcfs after GenotypeGVCFs if the intervals don't overlap. We typically keep them separate, but I've used GatherVcfs in the past with a large number of samples.

**Comment by droazen** [comment]
@mlathara and/or @nalinigans , what are your current recommendations for parallelizing a GenomicsDB-based pipeline?

**Comment by nalinigans** [comment]
The only parallelism exposed by GenomicsDBImport is via `max-num-intervals-to-import-in-parallel` and that is at the granularity of threads in the same process. Not sure there is any parallelism in GenotypeGVCFs. 
The granularity of parallelism that @nickhsmith seems to want is at the level of a node and @mjohnsonngi is correct. Assuming GenotypeGVCFs processes each position independently, run GenomicsDBImport followed by GenotypeGVFs per node and then use GatherVcfs.

**Comment by wichadak** [comment]
Hi, if I have 700 samples, can I use batch-size=700?

**Comment by mjohnsonngi** [comment]
> Hi, if I have 700 samples, can I use batch-size=700?

You could, but using 0 for batch size has the same effect of not using batches. It's the default, so not including --batch-size should have the same effect.

**Comment by igibmukeshkumar** [comment]
Hi @mjohnsonngi,

I am currently performing joint genotyping for 242 whole-exome samples. Given the sample size, I am following GATK best practices and using GenomicsDBImport to combine the individual GVCF files.

For this task, I am utilising a dedicated node with 380 GB RAM and 40 CPU cores.

Below are the parameters I am using to run GenomicsDBImport:

```

#!/bin/bash
#SBATCH -N 1                    # Number of nodes
#SBATCH -n 40                   # Number of tasks
#SBATCH -c 1                    # Number of CPU cores per task
#SBATCH -p patition_1              
#SBATCH -J v2              
#SBATCH --output=v2_%j_out.log   
#SBATCH --error=v2_%j_err.log   
#SBATCH --mail-type=ALL         
#SBATCH --mail-user=sainimuk77@gmail.com 

# Run GenomicsDBImport
gatk --java-options "-Xmx32g" GenomicsDBImport \
    --genomicsdb-workspace-path GT_cohort/geno_db \
    --sample-name-map sample_map.txt \
    --reader-threads 8 \
    --batch-size 50 \
    -L intervals.bed \
    --merge-input-intervals

```

**Log File:**

Using GATK jar /lustre/binukumar/tools/gatk-4.3.0.0/gatk-package-4.3.0.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xmx32g -jar /lustre/user/tools/gatk-4.3.0.0/gatk-package-4.3.0.0-local.jar GenomicsDBImport --genomicsdb-workspace-path GT_cohort/geno_db --sample-name-map sample_map.txt --reader-threads 8 --batch-size 50 -L intervals.bed --merge-input-intervals
13:20:48.538 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/lustre/user/tools/gatk-4.3.0.0/gatk-package-4.3.0.0-local.jar!/com/intel/gkl/native/libgkl_compression.so
13:20:48.847 INFO  GenomicsDBImport - ------------------------------------------------------------
13:20:48.847 INFO  GenomicsDBImport - The Genome Analysis Toolkit (GATK) v4.3.0.0
13:20:48.847 INFO  GenomicsDBImport - For support and documentation go to https://software.broadinstitute.org/gatk/
13:20:48.847 INFO  GenomicsDBImport - Executing as binukumar@node16 on Linux v3.10.0-1062.el7.x86_64 amd64
13:20:48.847 INFO  GenomicsDBImport - Java runtime: OpenJDK 64-Bit Server VM v11.0.13+7-b1751.21
13:20:48.847 INFO  GenomicsDBImport - Start Date/Time: May 19, 2025 at 1:20:48 PM IST
13:20:48.848 INFO  GenomicsDBImport - ------------------------------------------------------------
13:20:48.848 INFO  GenomicsDBImport - ------------------------------------------------------------
13:20:48.848 INFO  GenomicsDBImport - HTSJDK Version: 3.0.1
13:20:48.848 INFO  GenomicsDBImport - Picard Version: 2.27.5
13:20:48.848 INFO  GenomicsDBImport - Built for Spark Version: 2.4.5
13:20:48.848 INFO  GenomicsDBImport - HTSJDK Defaults.COMPRESSION_LEVEL : 2
13:20:48.848 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false
13:20:48.849 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true
13:20:48.849 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false
13:20:48.849 INFO  GenomicsDBImport - Deflater: IntelDeflater
13:20:48.849 INFO  GenomicsDBImport - Inflater: IntelInflater
13:20:48.849 INFO  GenomicsDBImport - GCS max retries/reopens: 20
13:20:48.849 INFO  GenomicsDBImport - Requester pays: disabled
13:20:48.849 INFO  GenomicsDBImport - Initializing engine
13:20:49.662 INFO  FeatureManager - Using codec BEDCodec to read file file:///intervals.bed
13:20:50.090 INFO  IntervalArgumentCollection - Processing 56159770 bp from intervals
13:20:50.194 INFO  GenomicsDBImport - Done initializing engine
13:20:50.411 INFO  GenomicsDBLibLoader - GenomicsDB native library version : 1.4.3-6069e4a
13:20:50.414 INFO  GenomicsDBImport - Vid Map JSON file will be written to GT_cohort/geno_db/vidmap.json
13:20:50.414 INFO  GenomicsDBImport - Callset Map JSON file will be written to GT_cohort/geno_db/callset.json
13:20:50.414 INFO  GenomicsDBImport - Complete VCF Header will be written to GT_cohort/geno_db/vcfheader.vcf
13:20:50.414 INFO  GenomicsDBImport - Importing to workspace - GT_cohort/geno_db
13:20:50.414 WARN  GenomicsDBImport - GenomicsDBImport cannot use multiple VCF reader threads for initialization when the number of intervals is greater than 1. Falling back to serial VCF reader initialization.
13:21:06.200 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
14:56:36.078 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
16:15:37.373 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
17:17:13.118 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
18:06:14.304 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
18:56:52.274 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
19:47:00.509 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
20:39:59.457 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
21:20:02.326 INFO  GenomicsDBImport - Importing batch 1 with 50 samples

....................................................................................................................................
....................................................................................................................................
....................................................................................................................................
06:23:08.916 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
06:23:15.351 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
06:23:21.773 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
06:23:28.466 INFO  GenomicsDBImport - Importing batch 1 with 50 samples
06:23:31.742 INFO  GenomicsDBImport - Done importing batch 1/5
06:23:44.317 WARN  IntelInflater - Zero Bytes Written : 0
06:23:45.369 INFO  GenomicsDBImport - Importing batch 2 with 50 samples
07:40:52.693 WARN  IntelInflater - Zero Bytes Written : 0
07:40:53.837 INFO  GenomicsDBImport - Importing batch 2 with 50 samples
08:42:27.617 WARN  IntelInflater - Zero Bytes Written : 0
08:42:27.907 INFO  GenomicsDBImport - Importing batch 2 with 50 samples
09:30:06.657 WARN  IntelInflater - Zero Bytes Written : 0
09:30:06.908 INFO  GenomicsDBImport - Importing batch 2 with 50 samples
10:06:34.669 WARN  IntelInflater - Zero Bytes Written : 0
10:06:34.936 INFO  GenomicsDBImport - Importing batch 2 with 50 samples
10:45:10.455 WARN  IntelInflater - Zero Bytes Written : 0
10:45:11.023 INFO  GenomicsDBImport - Importing batch 2 with 50 samples
11:23:04.846 WARN  IntelInflater - Zero Bytes Written : 0
11:23:05.156 INFO  GenomicsDBImport - Importing batch 2 with 50 samples
12:05:35.745 WARN  IntelInflater - Zero Bytes Written : 0
12:05:36.019 INFO  GenomicsDBImport - Importing batch 2 with 50 samples
12:36:33.516 WARN  IntelInflater - Zero Bytes Written : 0
12:36:33.855 INFO  GenomicsDBImport - Importing batch 2 with 50 samples

**Problem:**

I’m running joint genotyping on 242 whole-exome samples using GenomicsDBImport, but the job has been running for over 24 hours, and only one batch (batch-size = 50) has completed so far.

I’m using a BED file based on the Twist Comprehensive Exome kit, which I merged using bedtools.

The log shows a warning about intervals being more than 100, which prevents multithreading and slows down the import significantly.

Since exome BED files naturally contain a large number of intervals, how should I proceed to efficiently run GenomicsDBImport in such a scenario?

This is becoming a bottleneck due to the long runtime.

Any suggestions or best practices to address this would be greatly appreciated.

Thanks,
Mukesh
